ext/nnstreamer/tensor_filter/tensor_filter_trix_engine.cc

```diff
@@ -0,0 +1,353 @@
+      return DATA_LAYOUT_MODEL;
+  }
+}
+
+/**
+ * @brief Convert data type (from NNStreamer to TRIx-Engine)
+ */
+data_type
+TensorFilterTRIxEngine::convert_data_type (const tensor_type &type) {
+  switch (type) {
+    case _NNS_INT32:
+      return DATA_TYPE_INT32;
+    case _NNS_UINT32:
+      return DATA_TYPE_UINT32;
+    case _NNS_INT16:
+      return DATA_TYPE_INT16;
+    case _NNS_UINT16:
+      return DATA_TYPE_UINT16;
+    case _NNS_INT8:
+      return DATA_TYPE_INT8;
+    case _NNS_UINT8:
+      return DATA_TYPE_UINT8;
+    case _NNS_FLOAT64:
+      return DATA_TYPE_FLOAT64;
+    case _NNS_FLOAT32:
+      return DATA_TYPE_FLOAT32;
+    case _NNS_INT64:
+      return DATA_TYPE_INT64;
+    case _NNS_UINT64:
+      return DATA_TYPE_UINT64;
+    default:
+      return DATA_TYPE_MODEL;
+  }
+}
+
+/**
+ * @brief Set data info of input/output tensors
+ */
+void
+TensorFilterTRIxEngine::set_data_info (tensors_data_info *in_info, tensors_data_info *out_info) {
+  in_info->num_info = model_meta_->input_seg_num;
+
+  for (uint32_t idx = 0; idx < in_info->num_info; ++idx) {
+    in_info->info[idx].layout = convert_data_layout (input_layout_[idx]);
+    in_info->info[idx].type = convert_data_type (input_info_.info[idx].type);
+  }
+
+  out_info->num_info = model_meta_->output_seg_num;
+
+  for (uint32_t idx = 0; idx < out_info->num_info; ++idx) {
+    out_info->info[idx].layout = convert_data_layout (output_layout_[idx]);
+    out_info->info[idx].type = convert_data_type (output_info_.info[idx].type);
+  }
+}
+
+/**
+ * @brief Feed the tensor data to input buffers
+ */
+void
+TensorFilterTRIxEngine::feed_input_data (const GstTensorMemory *input, input_buffers *input_buf) {
+  input_buf->num_buffers = model_meta_->input_seg_num;
+
+  for (uint32_t idx = 0; idx < input_buf->num_buffers; ++idx) {
+    input_buf->bufs[idx].addr = input[idx].data;
+    input_buf->bufs[idx].size = input[idx].size;
+    input_buf->bufs[idx].type = BUFFER_MAPPED;
+  }
+}
+
+/**
+ * @brief Extract the tensor data from output buffers
+ */
+void
+TensorFilterTRIxEngine::extract_output_data (const output_buffers *output_buf,
+                                             GstTensorMemory *output) {
+  /* internal logic error */
+  assert (output_buf->num_buffers == model_meta_->output_seg_num);
+
+  for (uint32_t idx = 0; idx < output_buf->num_buffers; ++idx) {
+    output[idx].data = output_buf->bufs[idx].addr;
+    output[idx].size = output_buf->bufs[idx].size;
+  }
+}
+
+/**
+ * @brief Invoke TRIxEngine using input tensors
+ */
+void
+TensorFilterTRIxEngine::invoke (const GstTensorMemory *input, GstTensorMemory *output) {
+  int req_id;
+  int status;
+
+  status = createNPU_request (dev_, model_id_, &req_id);
+  if (status != 0) {
+    nns_loge ("Unable to create NPU request with model id (%u): %d", model_id_, status);
+    return;
+  }
+
+  tensors_data_info in_info;
+  tensors_data_info out_info;
+  /* set data info using metadata */
+  set_data_info (&in_info, &out_info);
```

Comment:
how about calling set_data_info() in configure_instance()?
(define member val in_info_ / out_info_ and remove tensor_layout) 
IMO, converting tensors info can be done in init state.
```
private:
  tensors_data_info in_info_;
  tensors_data_info out_info_;

//below layout val is unnecessary. 
  const tensor_layout *input_layout_;  /**< The data layout of input tensors */ 
  const tensor_layout *output_layout_; /**< The data layout of output tensors */

void set_data_info (const GstTensorFilterProperties *prop)

```

Reply:
Looks better, thanks.